### DISCUSSION TITLE
migrate from cocos2dx 3.17 to axmol [win32]

### DISCUSSION THREAD

**Discussion by NgVThangBz** [axmolengine/axmol#1341] [Q&A]
I'm trying to migrate a cocos 3.17 project to axmol 2.0. I have moved all the code files to the axmol project, I have changed the namespace as well as the file name accordingly. It seems like my code is fine but when building the project I get error C3130 (Error C3130 Internal Compiler Error: failed to write block of code inserted into PDB (compile source file...) HelloCpp C:\GitHub\axmol\core\2d\ActionInstant.h
I'm trying win32, and next will be ios and android
Has anyone encountered this error?

**Comment by halx99** [comment]
Which your vs version, ensure `-Thost=x64` for cmake generate and build target arch x64 is preferred

**Reply by NgVThangBz** [reply]
i using axmol 2.0,
How setting "-Thost=x64",
and yes, im build x4

**Reply by halx99** [reply]
and visual studio version?

**Reply by NgVThangBz** [reply]
VS 2022

**Reply by halx99** [reply]
do you clean the old project build intermediate files?

**Reply by NgVThangBz** [reply]
I just moved the .h and .cpp files in the Classes folder to the Source folder in axmol project

**Reply by halx99** [reply]
looks like it's the visual studio cl compiler bug, can you try clang?

**Reply by halx99** [reply]
install llvm 16.0.6 from https://github.com/llvm/llvm-project/releases add to system env PATH, then

axmol build -p win32 -cc clang

**Reply by NgVThangBz** [reply]
![image_2023-09-20_16-29-08](https://github.com/axmolengine/axmol/assets/125545428/e2210e57-baa9-4617-9042-e71187895812)
im do it, and build error

**Reply by rh101** [reply]
Something may not be right with your Visual Studio installation. Run Visual Studio Installer, then make sure you have these ticked:
![image](https://github.com/axmolengine/axmol/assets/8603230/cf0a0203-1a30-4e1b-a9c9-ff05c5564f69)

**Reply by halx99** [reply]
The llvm not install or not in system env `PATH`, note: llvm require install manually

**Reply by NgVThangBz** [reply]
![image_2023-09-20_17-02-30](https://github.com/axmolengine/axmol/assets/125545428/a4d85fd6-2c95-40cf-9087-2799c5ae5486)
![image_2023-09-20_17-03-12](https://github.com/axmolengine/axmol/assets/125545428/924774e9-af16-4bce-a23d-07bc7d5e6ab5)
![image_2023-09-20_17-51-33](https://github.com/axmolengine/axmol/assets/125545428/7eb51ce8-29e3-4ea4-adad-4850f5179f09)
im do it, but not have file .sln

**Reply by halx99** [reply]
llvm no .sln, just build on command line, does the bin/**/xxx.exe produced after `axmol build -p win32 -cc clang` ?

**Reply by rh101** [reply]
What you're displaying is not possible with the commands I mentioned above; there shouldn't be `*.ninja` files in there.

It's best that you post all the commands you're using in your terminal to generate the above.  Actually, even better if you screenshot the terminal so we can see exactly what it is you're typing and the output from those commands.

**Reply by NgVThangBz** [reply]
@halx99 dont have .exe
what I used initially was:
'axmol new -p org.axmol.hellocpp -d C:\GitHub -l cpp --portrait HelloCpp'
'cd HelloCpp'
'axmol build -p win32'

**Comment by rh101** [comment]
@NgVThangBz How did you generate your VS2022 project files? State the specific cmake command you used to do so.

**Reply by NgVThangBz** [reply]
im use 'axmol build -p win32'

**Reply by rh101** [reply]
Open a command prompt, navigate to the root of your project folder, then try this:

`cmake -S . -B build_win32 -G "Visual Studio 17 2022" -A Win32`

That will generate the VS solution/project files in a folder named `build_win32`.  Once it finishes, go into the `build_win32` folder, and open up the generated `[projectname].sln` file in VS2022.  Build it via the VS2022 IDE, and see if that works.

**Reply by NgVThangBz** [reply]
i do it, and same error

**Reply by rh101** [reply]
Does the 64 bit version work?

If you haven't tried it, run this command:
`cmake -S . -B build_x64 -G "Visual Studio 17 2022" -A x64`

Then open the VS solution in the directory `build_x64`, and build.

**Reply by halx99** [reply]
> im use 'axmol build -p win32'

default arch is 'x64' if not specify by `-a x86`

**Reply by NgVThangBz** [reply]
@halx99 
i tried  'axmol build -p win32 -a x86' but got the same error
is there still any problem if i port my code from my cocos2dx 3.17 project over to axmol
![image_2023-09-21_09-05-36](https://github.com/axmolengine/axmol/assets/125545428/a88f3b78-b548-4665-bd03-0c2860229c36)

**Reply by halx99** [reply]
as i said, it's looks like msvc compiler bug, *not related to build arch*, it's difficult to fire which code cause the error. so try update visual studio or use llvm-clang

**Reply by thienphuoc** [reply]
![image](https://github.com/user-attachments/assets/a1b49fc2-b15f-4525-9cb6-b6a756d75eaf)
After you have installed ClangCL, specify that the project should be built using ClangCL by running the following CMake command:
``
cmake -S . -B build -G "Visual Studio 17 2022" -A Win32 -T ClangCL
``
To confirm that ClangCL is being used, open the project in Visual Studio and navigate to:

Configuration Properties → General → Platform Toolset

If you see LLVM (clang-cl) as the selected toolset, then everything is set up correctly.

![image](https://github.com/user-attachments/assets/ddbbc42d-97ed-44e6-a2fc-77b0ad2c1378)

**Comment by NgVThangBz** [comment]
i setup: Properties-> C/C++ -> Debug Info: C7 (Z7)
and now it work
